ocaml-lsp-server/src/diagnostics.mli

```diff
@@ -0,0 +1,16 @@
+open Import
+
+type t
+
+val create : (PublishDiagnosticsParams.t -> unit Fiber.t) -> t
+
+val send : t -> unit Fiber.t
+
+val set :
+     t
+  -> [ `Dune of int * Uri.t * Diagnostic.t
+     | `Merlin of Uri.t * Diagnostic.t list
```

Comment:
Should you know where the diagnostic come from? Why not `val set : t -> (Ur.t * Diagnostic.t) list -> unit`

Also, maybe an immutable API would be safer: 

```
val set : t -> (Ur.t * Diagnostic.t) list -> t
```

I'm not sure if this would become a bottleneck performance-wise.

Reply:
> Why not val set : t -> (Ur.t * Diagnostic.t) list -> unit

Because dune does not send us diagnostics per document. It sends add/remove events in arbitrary order. This is why I need the hash table keyed by ID to maintain the current set of diagnostics efficiently.

> I'm not sure if this would become a bottleneck performance-wise.

I'm not worried about the performance but the about boilerplate overhead. This can be experimented with however.